### ISSUE TITLE
Unicode URL support

### ISSUE THREAD

**Issue by dantheta** [openrightsgroup/cmp-issues#122]
While I was checking over the DMOZ import, I noticed a set of URLs that had failed to import because they contained unicode characters in the path.  Although we don't currently have any support for internationalized domain names, the path components can be specified in any character set.  Our URLs table is currently set up for latin1 URLs only.  The DMOZ import used a latin1 database connection, which resulted in minor character corruption in category names (this is being fixed).

Do we want to support unicode URLs at this stage?

**Comment by graphiclunarkid** [comment]
I think we do at some point - we have ambitions to deploy the software outside the UK so international language support would be very useful. Would it be a great deal of work though?

**Comment by dantheta** [comment]
It's pretty straightforward - it does reduce the maximum length of URL that the system can handle though, since our current version of MySQL will only uniquely index 255 characters of UTF8 data (we currently uniquely index 767 chars of latin1).  We'd also need to do a small amount of work to make sure that we're using the multibyte string functions in PHP any time we're manipulating URLs.

**Comment by dantheta** [comment]
I've added a feature branch for unicode URL support . 
https://github.com/openrightsgroup/Blocking-Middleware/tree/feature/unicode_urls

 There are database changes to improve the URL indexing - the URL itself is parsed and split into scheme, domain and path+query, giving us 510 unicode characters for the unique index.  The domain part is unicode-lowercased (addressing one of @gwire's previous feature requests surrounding URL normalization).  The system correctly converts the unicode domains into IDN punycode encoding behind the scenes.  

The supporting scripts and the robots.txt checker have also been updated.

I don't think we should roll it out to UK blocked straightaway, but it's a useful basis for international handling.
